enhancements/cluster-logging/forward-to-loki.md

```diff
@@ -0,0 +1,235 @@
+
+The benefit is faster ingest, which reduces the risk of dropping logs.
+The trade-off is that need to carefully choose what to use as labels.
+
+For more details:
+- [Guide to labels in loki](https://grafana.com/blog/2020/08/27/the-concise-guide-to-labels-in-loki/)
+- [Out-of-order streams issue]([https://github.com/grafana/loki/issues/1544)
+
+To summarize the key points:
+- Each unique combination of labels defines an ordered Loki _stream_.
+  - Too many streams degrades performance.
+  - _Cardinality_ refers to the number of unique label combinations.
+- The cardinality of `kubernetes.pod_name` was found to be
+  - too high in clusters with _very_ rapid turn-over of short-lived pods.
+  - acceptable in other (possibly more typical) clusters with high log data rates.
+- Minimize the _total number_ of labels per stream
+  - There is a limit of 30 max, fewer labels == better performance.
+- Log streams must be _ordered by collection timestamp_.
+  - Must not allow distinct nodes with separate clocks to contribute to a single stream.
+
+All logging meta-data will still be included in log records.
+Labels do not need to _identify the source of logs_, only to _partition the search space_.
+At query time labels reduce the search space, then Loki uses log content to complete the search.
+
+**Note**: Loki [label names][] must match the regex `[a-zA-Z_:][a-zA-Z0-9_:]*`.
+Meta-data names are translated to Loki labels by replacing illegal characters with '_'.
+
+Nested JSON objects are referenced by converting their JSON path to a label name.
+For example, if `kubernetes.namespace_name` is a label, then this JSON log record:
+
+``` json
+{"kubernetes.namespace_name":"foo","kubernetes":{"labels":{"foo":"bar"}}
+```
+would match this LogQL filter: 
+
+``` logql
+{ kubernetes_namespace_name="foo" } | json | kubernetes_labels_foo == "bar"
+```
+### Default Loki Labels
+
+The proposed default label set is:
+
+- `log_type`: one of "application", "infrastructure", "audit"
+- `cluster`: for muti-cluster deployments (see Open Questions).
+- `kubernetes.host`: aka node, ensures ordered streams.
+- `kubernetes.namespace_name`: likely to be most common known factor in a search.
+- `kubernetes.container_name`: low cardinality, usually well-known
+
+`kubernetes.pod_name` is _not_ included. It is potentially high cardinality.
+`kubernetes.namespace_name` can be used to constrain the search for a named pod.
+
+The API below allows the user to add labels that are relevant for their application.
+
+### Proposed API
+
+Add a new `loki` output type to the `ClusterLogForwarder` API:
+
+``` yaml
+- name: myLokiOutput
+  type: loki
+  url: ...
+  secret: ...
+```
+
+`url` and `secret` have the usual meaning with regard to TLS and certificates.
+The `secret` may also contain `username` and `password` fields for Loki.
+
+The following optional fields are Loki-specific:
+
+- `labelKeys`: ([]string, optional)  A list of meta-data keys to add to the label set.\
```

Comment:
@alanconway taking into account that we do not know exactly the correct defaults etc ... why there is just an option to "add" keys to the list ... maybe we also need to support "remove" default keys from the list ... actually ,... you can change the definition of "lablekeys" to overwrite the list of kesy to custom list and allow users to do whatever they want with the keys

Reply:
@eranra I believe the schema here is:
- The Cluster Logging stack decides a bare minimum set of labels to be always present to support use cases in the OCP console, e.g. pod detail log view.
- The user can add more labels to aggregate them in the index if he desires so. Removing labels from his/her configuration is as simple as dropping them from `labelKeys` again.

Reply:
@periklis hence my point ... do we really need to force customers to that required list of labels ... if so, the list should really be bare minimum set of labels... and I argue that we want to set as default the recommended set of labels and not minimum required … so we want to allow customers to change labels that are recommended and not required

Reply:
@eranra I don't believe we will ever expose our minimum set of labels in `labelKeys`. Latter is only to add additional labels from scraped kubernetes labels field. We will certainly not recommend anything here except the pod name.

Reply:
@eranra  I have re-worded so that the user can _replace_ the default label set (e.g. not include pod name, or use UIDs instead of names) with the exception of the host name, which is mandatory to ensure ordered streams.